### PR TITLE
Skip alpn rewrite if alpnOverride filter metadata is false

### PR DESCRIPTION
**Pull request by ksubrmnn**
**What this PR does / why we need it**:
Skips alpn header rewrite if alpnOverride filter metadata is false. This metadata is set by Istio when TLS mode is simple.

Works with https://github.com/istio/istio/pull/44918

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #
Partially solves https://github.com/istio/istio/issues/40680

**Special notes for your reviewer**:
Please read the [RFC](https://docs.google.com/document/d/1T8NvwK2hnRjU-LWWtZ_Wk6kIVkeZqYEcP_LQ3dEkO-8/edit) for the design details

Can I get a pointer for where/how to add a test?

### CODE REVIEW

**Comment by ksubrmnn** [comment]
cc @howardjohn @keithmattix

**Review by howardjohn — COMMENTED** [review]
conceptually lgtm but I am not an expert on this code so I won't/can't approve

cc @kyessenov

File: source/extensions/filters/http/alpn/alpn_filter.cc [hunk 1]
```diff
@@ -55,6 +55,24 @@ Http::Protocol AlpnFilterConfig::getHttpProtocol(
 }
 
 Http::FilterHeadersStatus AlpnFilter::decodeHeaders(Http::RequestHeaderMap&, bool) {
+  auto upstream_info = decoder_callbacks_->streamInfo().upstreamInfo();
```
Comment (kyessenov) [inline source/extensions/filters/http/alpn/alpn_filter.cc:58]: nit: const?

**Review by kyessenov — COMMENTED** [review]
Looks fine, can you add a test?

**Comment by kyessenov** [comment]
asan is a flake, you can ignore stackdriver test failures
/retest

**Comment by keithmattix** [comment]
Part of istio/istio#40680

**Review by howardjohn — APPROVED** [review]

File: source/extensions/filters/http/alpn/alpn_test.cc [hunk 2]
```diff
@@ -167,6 +167,35 @@ TEST_F(AlpnFilterTest, EmptyOverrideAlpn) {
   }
 }
 
+TEST_F(AlpnFilterTest, AlpnOverrideFalse) {
+  NiceMock<StreamInfo::MockStreamInfo> stream_info;
+  NiceMock<StreamInfo::MockUpstreamInfo> upstream_info;
+  Envoy::Upstream::HostDescriptionConstSharedPtr upstream_host; // TODO: generate mock?, remove pointer
+  Envoy::Upstream::MetadataConstSharedPtr host_metadata; // TODO: generate mock?, remove pointer
```
Comment (jewertow) [inline source/extensions/filters/http/alpn/alpn_test.cc:174]: You don't have to generate mocks. You can apply this commit to fix this test: https://github.com/istio/proxy/pull/4847/commits/895c7c5f19449259eeb3f7afb90f468ed043ee9d

Comment (jewertow) [reply source/extensions/filters/http/alpn/alpn_test.cc:174]: You can see in [this PR](https://github.com/istio/proxy/pull/4847) that tests pass with this commit.

**Review by hzxuzhonghu — APPROVED** [review]

**Review by zirain — APPROVED** [review]

**Pull request merged**